tsdb/engine/wal/wal.go

```diff
@@ -0,0 +1,1013 @@
+func (p *partition) segmentFileNames() ([]string, error) {
+	path := filepath.Join(p.path, fmt.Sprintf("%02d.*.%s", p.id, FileExtension))
+	fileNames, err := filepath.Glob(path)
+	return fileNames, err
+}
+
+// segmentFile is a struct for reading in segment files from the WAL. Used on startup only while loading
+type segmentFile struct {
+	f      *os.File
+	block  []byte
+	length []byte
+	size   int
+}
+
+func newSegmentFile(f *os.File) *segmentFile {
+	return &segmentFile{
+		length: make([]byte, 8),
+		f:      f,
+	}
+}
+
+// readCompressedBlock will read the next compressed block from the file and marshal the entries.
+// if we've hit the end of the file or corruption the entry array will be nil
+func (s *segmentFile) readCompressedBlock() (entries []*entry, err error) {
+	n, err := s.f.Read(s.length)
+	if err == io.EOF {
+		return nil, nil
+	} else if err != nil {
+		return nil, err
+	}
+	s.size += n
+
+	if n != len(s.length) {
+		log.Println("unable to read the size of a data block from file: ", s.f.Name())
+		// seek back before this length so we can start overwriting the file from here
+		s.f.Seek(-int64(n), 1)
+		return nil, nil
+	}
+
+	// Compacted WAL files will have a magic byte sequence that indicate the next part is a file name
+	// instead of a compressed block. We can ignore these bytes and the ensuing file name to get to the next block.
+	isCompactionFileNameBlock := false
+	if bytes.Compare(s.length[0:2], CompactSequence) == 0 {
+		s.length[0] = 0x00
+		s.length[1] = 0x00
+		isCompactionFileNameBlock = true
+	}
+
+	dataLength := btou64(s.length)
+
+	// make sure we haven't hit the end of data. trailing end of file can be zero bytes
+	if dataLength == 0 {
+		s.f.Seek(-int64(len(s.length)), 1)
+		return nil, nil
+	}
+
+	if len(s.block) < int(dataLength) {
+		s.block = make([]byte, dataLength)
+	}
+
+	n, err = s.f.Read(s.block[:dataLength])
+	if err != nil {
+		return nil, err
+	}
+	s.size += n
+
+	// read the compressed block and decompress it. if partial or corrupt,
+	// overwrite with zeroes so we can start over on this wal file
+	if n != int(dataLength) {
+		log.Println("partial compressed block in file: ", s.f.Name())
+
+		// seek back to before this block and its size so we can overwrite the corrupt data
+		s.f.Seek(-int64(len(s.length)+n), 1)
+		if err := s.zeroRestOfFile(); err != nil {
+			return nil, err
+		}
+
+		return nil, nil
+	}
+
+	// skip the rest if this is just the filename from a compaction
+	if isCompactionFileNameBlock {
+		return nil, ErrCompactionBlock
+	}
+
+	buf, err := snappy.Decode(nil, s.block[:dataLength])
+
+	// if there was an error decoding, this is a corrupt block so we zero out the rest of the file
+	if err != nil {
+		log.Println("corrupt compressed block in file: ", err.Error(), s.f.Name())
+
+		// go back to the start of this block and zero out the rest of the file
+		s.f.Seek(-int64(len(s.length)+n), 1)
+		if err := s.zeroRestOfFile(); err != nil {
+			return nil, err
+		}
```

Comment:
I wanted to ensure that the file wasn't rewritten by the OS. Guess that's something I probably shouldn't be worrying about at this point though. Truncate would definitely be easier.